### PR TITLE
Provide `--platform` / `--subdir` flag to create (and remember) an environment for a platform different than native

### LINKED ISSUE #11505: Set conda platform in environment creation subcommands and record it in environment config
### Checklist

- [X] I added a descriptive title
- [X] I searched for other feature requests and couldn't find a duplicate (including also the ``type-feature`` tag)
- [X] I confirmed that it's not related to another project (see the above section)

### What is the idea?

An argument (maybe `--platform`?) added to all conda subcommands that create environments which sets the platform subdirectory to something other than the default platform and also records the platform in the environment-specific configuration so it will be used in subsequent commands when the environment is active.

### Why is this needed

The M1 Macs offer a novel situation where the OS supports two different conda platforms: `osx-64` and `osx-arm64`.  As not all software has been ported to the M1, it is frequently useful to create an x86 environment when the base environment is ARM, or vice versa.  The current workaround is either:
* Set the `CONDA_SUBDIR` environment whenever running conda.  However, if you forget on a subsequent update, you could mess up the environment.
* Create an empty environment and set the environment config.  Example:
```
# starting from osx-64 base environment
conda create -n m1 
conda activate m1
# set the platform in just this environment (--env) to osx-arm64
conda config --env --set subdir osx-arm64
# all future installs and updates will use osx-arm64 packages
conda install python=3.8 notebook pandas
```

I can imagine future use cases where this could be useful (say, for conda managing a micropython environment), but those are speculative.  The immediate driving use case is for M1 Macs.

### What should happen?

I would like to run the command:
```
conda create --platform osx-64 -n x86_test python=3.7 notebook
```
and have the new environment populated with x86 packages, and also have the platform preference recorded in the conda environment-specific configuration file.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jaimergp**
### Description

With the release of M1 processors in Apple products, a common question is "how do I create an Intel environment in my M1 laptop?". This is technically possible today but involves two (obscure) steps:

1. Set `CONDA_SUBDIR` to trick `conda` into thinking we are in a different platform.
2. Make `conda` remember that choice _after_ environment creation and activation via `conda config --env --set subdir <platform>`.

This PR simplifies that setup into a single step: `conda create` and `conda env create` will now accept a `--platform` CLI flag that automates the process.

This follows these precedents:

* `conda search` supports a `--platform` / `--subdir` flag to find packages in other platforms
* `micromamba create` [supports](https://github.com/mamba-org/mamba/blob/0cc361562a3d0c712d042f1c10c790eb8133a366/libmamba/src/api/create.cpp#L79-L90) a `--platform` flag, the value of which will get saved to `$PREFIX/.mambarc`.

This is how the `conda create` CLI help would look like:

```
$ conda create --help
usage: conda create [-h] [--clone ENV] (-n ENVIRONMENT | -p PATH) [-c CHANNEL] [--use-local]
                    [--override-channels] [--repodata-fn REPODATA_FNS] [--strict-channel-priority]
                    [--no-channel-priority] [--no-deps | --only-deps] [--no-pin] [--copy] [-C] [-k]
                    [--offline] [-d] [--json] [-q] [-v] [-y] [--download-only]
                    [--show-channel-urls] [--file FILE] [--no-default-packages]
                    [--experimental-solver {classic,libmamba,libmamba-draft}] [--subdir SUBDIR]
                    [--dev]
                    [package_spec [package_spec ...]]

Create a new conda environment from a list of specified packages. To use the created environment, use 'conda activate envname' to look in that directory first.  This command requires either the -n NAME or -p PREFIX option.

Options:

positional arguments:
  package_spec          Packages to install or update in the conda environment.

optional arguments:
  -h, --help            Show this help message and exit.
  --clone ENV           Path to (or name of) existing local environment.
  --file FILE           Read package versions from the given file. Repeated file specifications can
                        be passed (e.g. --file=file1 --file=file2).
  --dev                 Use `sys.executable -m conda` in wrapper scripts instead of CONDA_EXE. This
                        is mainly for use during tests where we test new conda source against old
                        Python versions.

Target Environment Specification:
  -n ENVIRONMENT, --name ENVIRONMENT
                        Name of environment.
  -p PATH, --prefix PATH
                        Full path to environment location (i.e. prefix).

Channel Customization:
  -c CHANNEL, --channel CHANNEL
                        Additional channel to search for packages. These are URLs searched in the
                        order they are given (including local directories using the 'file://'
                        syntax or simply a path like '/home/conda/mychan' or '../mychan'). Then,
                        the defaults or channels from .condarc are searched (unless --override-
                        channels is given). You can use 'defaults' to get the default packages for
                        conda. You can also use any name and the .condarc channel_alias value will
                        be prepended. The default channel_alias is https://conda.anaconda.org/.
  --use-local           Use locally built packages. Identical to '-c local'.
  --override-channels   Do not search default or .condarc channels. Requires --channel.
  --repodata-fn REPODATA_FNS
                        Specify name of repodata on remote server. Conda will try whatever you
                        specify, but will ultimately fall back to repodata.json if your specs are
                        not satisfiable with what you specify here. This is used to employ repodata
                        that is reduced in time scope. You may pass this flag more than once.
                        Leftmost entries are tried first, and the fallback to repodata.json is
                        added for you automatically.
  --subdir SUBDIR, --platform SUBDIR
                        Use packages built for this platform. The new environment will be
                        configured to remember this choice. Should be formatted like 'osx-64',
                        'linux-32', 'win-64', and so on. Defaults to the current (native) platform.

Solver Mode Modifiers:
  --strict-channel-priority
                        Packages in lower priority channels are not considered if a package with
                        the same name appears in a higher priority channel.
  --no-channel-priority
                        Package version takes precedence over channel priority. Overrides the value
                        given by `conda config --show channel_priority`.
  --no-deps             Do not install, update, remove, or change dependencies. This WILL lead to
                        broken environments and inconsistent behavior. Use at your own risk.
  --only-deps           Only install dependencies.
  --no-pin              Ignore pinned file.
  --no-default-packages
                        Ignore create_default_packages in the .condarc file.
  --experimental-solver {classic,libmamba,libmamba-draft}
                        EXPERIMENTAL. Choose which solver backend to use.

Package Linking and Install-time Options:
  --copy                Install all packages using copies instead of hard- or soft-linking.

Networking Options:
  -C, --use-index-cache
                        Use cache of channel index files, even if it has expired.
  -k, --insecure        Allow conda to perform "insecure" SSL connections and transfers. Equivalent
                        to setting 'ssl_verify' to 'false'.
  --offline             Offline mode. Don't connect to the Internet.

Output, Prompt, and Flow Control Options:
  -d, --dry-run         Only display what would have been done.
  --json                Report all output as json. Suitable for using conda programmatically.
  -q, --quiet           Do not display progress bar.
  -v, --verbose         Can be used multiple times. Once for INFO, twice for DEBUG, three times for
                        TRACE.
  -y, --yes             Do not ask for confirmation.
  --download-only       Solve an environment and ensure package caches are populated, but exit
                        prior to unlinking and linking packages into the prefix.
  --show-channel-urls   Show channel urls. Overrides the value given by `conda config --show
                        show_channel_urls`.

Examples:

    conda create -n myenv sqlite
```

And for `conda env create`:

```
conda env create --help
usage: conda-env create [-h] [-f FILE] [-n ENVIRONMENT | -p PATH] [-C] [-k] [--offline] [--force]
                        [-d] [--no-default-packages] [--json] [-v] [-q]
                        [--experimental-solver {classic,libmamba,libmamba-draft}] [--subdir SUBDIR]
                        [remote_definition]

Create an environment based on an environment definition file.

If using an environment.yml file (the default), you can name the
environment in the first line of the file with 'name: envname' or
you can specify the environment name in the CLI command using the
-n/--name argument. The name specified in the CLI will override
the name specified in the environment.yml file.

Unless you are in the directory containing the environment definition
file, use -f to specify the file path of the environment definition
file you want to use.

Options:

positional arguments:
  remote_definition     Remote environment definition / IPython notebook

optional arguments:
  -h, --help            Show this help message and exit.
  -f FILE, --file FILE  Environment definition file (default: environment.yml)
  --force               Force creation of environment (removing a previously-existing environment
                        of the same name).
  -d, --dry-run         Only display what can be done with the current command, arguments, and
                        other flags. Remove this flag to actually run the command.
  --no-default-packages
                        Ignore create_default_packages in the .condarc file.
  --experimental-solver {classic,libmamba,libmamba-draft}
                        EXPERIMENTAL. Choose which solver backend to use.
  --subdir SUBDIR, --platform SUBDIR
                        Use packages built for this platform. The new environment will be
                        configured to remember this choice. Should be formatted like 'osx-64',
                        'linux-32', 'win-64', and so on. Defaults to the current (native) platform.

Target Environment Specification:
  -n ENVIRONMENT, --name ENVIRONMENT
                        Name of environment.
  -p PATH, --prefix PATH
                        Full path to environment location (i.e. prefix).

Networking Options:
  -C, --use-index-cache
                        Use cache of channel index files, even if it has expired.
  -k, --insecure        Allow conda to perform "insecure" SSL connections and transfers. Equivalent
                        to setting 'ssl_verify' to 'false'.
  --offline             Offline mode. Don't connect to the Internet.

Output, Prompt, and Flow Control Options:
  --json                Report all output as json. Suitable for using conda programmatically.
  -v, --verbose         Use once for info, twice for debug, three times for trace.
  -q, --quiet           Do not display progress bar.

examples:
    conda env create
    conda env create -n envname
    conda env create folder/envname
    conda env create -f /path/to/environment.yml
    conda env create -f /path/to/requirements.txt -n envname
    conda env create -f /path/to/requirements.txt -p /home/user/envname
```

Note that while `context.subdir` can be set to _anything_ via `CONDA_SUBDIR`, we do check for valid values in the CLI for user-friendliness. Compare the two errors:

```
$ conda create -n test-platform python --platform=osx-16
usage: conda create [-h] [--clone ENV] (-n ENVIRONMENT | -p PATH) [-c CHANNEL] [--use-local]
                    [--override-channels] [--repodata-fn REPODATA_FNS] [--strict-channel-priority]
                    [--no-channel-priority] [--no-deps | --only-deps] [--no-pin] [--copy] [-C] [-k]
                    [--offline] [-d] [--json] [-q] [-v] [-y] [--download-only]
                    [--show-channel-urls] [--file FILE] [--no-default-packages]
                    [--experimental-solver {classic,libmamba,libmamba-draft}] [--subdir SUBDIR]
                    [--dev]
                    [package_spec [package_spec ...]]
conda create: error: argument --subdir/--platform: invalid choice: 'osx-16' (choose from 'linux-32', 'linux-64', 'linux-aarch64', 'linux-armv6l', 'linux-armv7l', 'linux-ppc64', 'linux-ppc64le', 'linux-s390x', 'osx-64', 'osx-arm64', 'win-32', 'win-64', 'win-arm64', 'zos-z')
```

```
CONDA_SUBDIR=osx-16 conda create -n test-platform python
Collecting package metadata (current_repodata.json): failed
WARNING conda.exceptions:print_unexpected_error_report(1251): KeyError('pkgs_dirs')
Traceback (most recent call last):
  File "/opt/conda-src/conda/exceptions.py", line 1129, in __call__
    return func(*args, **kwargs)
  File "/opt/conda-src/conda/cli/main.py", line 86, in main_subshell
    exit_code = do_call(args, p)
  File "/opt/conda-src/conda/cli/conda_argparse.py", line 93, in do_call
    return getattr(module, func_name)(args, parser)
  File "/opt/conda-src/conda/notices/core.py", line 72, in wrapper
    return_value = func(*args, **kwargs)
  File "/opt/conda-src/conda/cli/main_create.py", line 43, in execute
    install(args, parser, 'create')
  File "/opt/conda-src/conda/cli/install.py", line 264, in install
    unlink_link_transaction = solver.solve_for_transaction(
  File "/opt/conda-src/conda/core/solve.py", line 156, in solve_for_transaction
    unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
  File "/opt/conda-src/conda/core/solve.py", line 199, in solve_for_diff
    final_precs = self.solve_final_state(update_modifier, deps_modifier, prune, ignore_pinned,
  File "/opt/conda-src/conda/core/solve.py", line 304, in solve_final_state
    ssc = self._collect_all_metadata(ssc)
  File "/opt/conda-src/conda/common/io.py", line 86, in decorated
    return f(*args, **kwds)
  File "/opt/conda-src/conda/core/solve.py", line 439, in _collect_all_metadata
    _supplement_index_with_system(virtual_pkg_index)
  File "/opt/conda-src/conda/core/index.py", line 215, in _supplement_index_with_system
    archspec_name = get_archspec_name()
  File "/opt/conda-src/conda/core/index.py", line 232, in get_archspec_name
    machine = _arch_names[int(target_arch)]
KeyError: 16

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda-src/conda/exceptions.py", line 1249, in print_unexpected_error_report
    message_builder.append(get_main_info_str(error_report['conda_info']))
  File "/opt/conda-src/conda/cli/main_info.py", line 237, in get_main_info_str
    info_dict['_' + key] = ('\n' + 26 * ' ').join(info_dict[key])
KeyError: 'pkgs_dirs'

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/opt/conda-src/conda/exceptions.py", line 1129, in __call__
        return func(*args, **kwargs)
      File "/opt/conda-src/conda/cli/main.py", line 86, in main_subshell
        exit_code = do_call(args, p)
      File "/opt/conda-src/conda/cli/conda_argparse.py", line 93, in do_call
        return getattr(module, func_name)(args, parser)
      File "/opt/conda-src/conda/notices/core.py", line 72, in wrapper
        return_value = func(*args, **kwargs)
      File "/opt/conda-src/conda/cli/main_create.py", line 43, in execute
        install(args, parser, 'create')
      File "/opt/conda-src/conda/cli/install.py", line 264, in install
        unlink_link_transaction = solver.solve_for_transaction(
      File "/opt/conda-src/conda/core/solve.py", line 156, in solve_for_transaction
        unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
      File "/opt/conda-src/conda/core/solve.py", line 199, in solve_for_diff
        final_precs = self.solve_final_state(update_modifier, deps_modifier, prune, ignore_pinned,
      File "/opt/conda-src/conda/core/solve.py", line 304, in solve_final_state
        ssc = self._collect_all_metadata(ssc)
      File "/opt/conda-src/conda/common/io.py", line 86, in decorated
        return f(*args, **kwds)
      File "/opt/conda-src/conda/core/solve.py", line 439, in _collect_all_metadata
        _supplement_index_with_system(virtual_pkg_index)
      File "/opt/conda-src/conda/core/index.py", line 215, in _supplement_index_with_system
        archspec_name = get_archspec_name()
      File "/opt/conda-src/conda/core/index.py", line 232, in get_archspec_name
        machine = _arch_names[int(target_arch)]
    KeyError: 16

`$ /opt/conda/bin/conda create -n test-platform python`

  environment variables:
conda info could not be constructed.
KeyError('pkgs_dirs')

An unexpected error has occurred. Conda has prepared the above report.

If submitted, this report will be used by core maintainers to improve
future releases of conda.
Would you like conda to send this report to the core maintainers? [y/N]: N

No report sent. To permanently opt-out, use

    $ conda config --set report_errors false
```

I left the context unvalidated (as it was) in case there's a technical reason somewhere where they need uber custom subdirs or something.

Closes #11505

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [X] Add / update necessary tests?
- [ ] Add / update outdated documentation?

### CODE REVIEW

File: conda/base/context.py [hunk 1]
```diff
@@ -492,6 +492,10 @@ def execute_threads(self):
     def subdir(self):
         if self._subdir:
             return self._subdir
+        return self._native_subdir()
+
+    @lru_cache(maxsize=None)
+    def _native_subdir(self):
```
Comment (jezdez) [inline conda/base/context.py:498]: Could we make this a `memoizedproperty` instead and remove the `_` prefix?

**Review by jezdez — DISMISSED** [review]
This is a good idea, but I think needs a documentation update.

File: conda/base/context.py (hunk 1 above)
Comment (dholth) [reply conda/base/context.py:498]: [```
        "The `conda.auxlib.decorators.memoize` decorator is pending deprecation and will be "
        "removed in a future release. Please use `functools.lru_cache` instead.",
```](https://github.com/conda/conda/blob/main/conda/auxlib/decorators.py#L53)

There may not be a working/non-deprecated function with a prettier name

Comment (jezdez) [reply conda/base/context.py:498]: Ah, I knew that, and forgot 😬 Ignore me!

**Comment by dholth** [comment]
I had weird side effects setting subdir and wound up installing x86 miniconda separately instead. We will have to carefully check the feature works as expected.

**Comment by jaimergp** [comment]
Yes, I think we need some kind of check to ensure the following situation does not happen:

* If you have a non-native environment, its `$CONDA_PREFIX/.condarc` will have `subdir: <non-native>`. 
* If you create a new environment with the non-native env ACTIVE, it will inherit the `subdir`, regardless whether you were using `--platform` or not, and the new environment will have that `subdir` too!

I see two solutions:
* Only allow non-native environments to be created from `base`.
* Carefully check where the `subdir` value is coming from and only allow it when the non-native value comes from either an environment variable or a command-line flag.

**Comment by pmeier** [comment]
Just dropping in to say that it would be really useful to have something like that. My use case is that part of our CI is unable to solve a particular environment. I want to be able to do something like `conda create --dry-run --platform=macos-64` in order to debug this locally and not go to the painstaking way of spinning up a new CI job and wait for signal every time. Meaning, for my use case, I don't even need the final env, but the solve needs to happen as if I were on a different platform.

**Comment by jaimergp** [comment]
> More work is needed

Can we discuss which option(s) [as described in this comment](https://github.com/conda/conda/pull/11794#issuecomment-1263405589) are a better idea?

**Review by jezdez — APPROVED** [review]

**Comment by jezdez** [comment]
Postponing again since the test failure seems legit.

**Pull request merged**